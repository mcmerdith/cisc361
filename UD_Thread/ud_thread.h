/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);
