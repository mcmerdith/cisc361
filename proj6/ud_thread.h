/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);

/**
 * semaphore
 */

// Create a new semaphore pointed to by sp with a count value of sem_count.
int sem_init(sem_t **sp, int sem_count);

// Current thread does a wait (P) on the specified semaphore.
void sem_wait(sem_t *sp);

// Current thread does a signal (V) on the specified semaphore.
// Follow the Mesa semantics (p. 9 of OSTEP Chapter 30 Condition Variables) where the thread that signals continues,
// and the first waiting (blocked) thread (if there is any) becomes ready.
void sem_signal(sem_t *sp);

// Free any memory related to the specified semaphore.
// If there are queued TCBs, they are moved to the ready queue, and this may imply that the semantics of the application might be incorrect.
void sem_destroy(sem_t **sp);

typedef void sem_t;