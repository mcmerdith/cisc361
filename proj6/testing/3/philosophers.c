#include "ud_thread.h"
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define TIME_TO_EAT 3
#define PHILOSOPHERS 5
static sem_t *forks[PHILOSOPHERS];
int eaten = 0;

sem_t *lock;
int forkInUse[PHILOSOPHERS];

void philosopher(int id)
{
    int right = id;
    int left = (id + 1) % PHILOSOPHERS;

    printf("Philosopher %d is hungry\n", id);
    if (id % 2 == 0)
    {
        sem_wait(forks[right]);
        sem_wait(lock);
        if (forkInUse[right])
        {
            printf("ERROR: fork %d is being used\n", right);
            exit(0);
        }
        forkInUse[right] = 1;
        sem_signal(lock);
        printf("Philosopher %d has picked up right fork %d\n", id, id);

        sem_wait(forks[left]);
        sem_wait(lock);
        if (forkInUse[left])
        {
            printf("ERROR: fork %d is being used\n", left);
            exit(0);
        }
        forkInUse[left] = 1;
        sem_signal(lock);
        printf("Philosopher %d has picked up left  fork %d\n", id, left);
    }
    else
    {
        sem_wait(forks[(id + 1) % PHILOSOPHERS]);
        printf("Philosopher %d has picked up left  fork %d\n", id, (id + 1) % PHILOSOPHERS);
        sem_wait(forks[id]);
        printf("Philosopher %d has picked up right fork %d\n", id, id);
    }

    int i = 0;
    for (; i < TIME_TO_EAT; i++)
    {
        t_yield();
        printf("Philosopher %d is still eating\n", id);
    }

    sem_signal(forks[id]);
    sem_signal(forks[(id + 1) % PHILOSOPHERS]);

    printf("Philosopher %d has eaten\n", id);
    sem_wait(lock);

    forkInUse[left] = 0;
    forkInUse[right] = 0;
    eaten++;
    sem_signal(lock);

    t_terminate();
}

int main()
{
    int i;
    srand(time(NULL));

    t_init();
    sem_init(&lock, 1);

    //  Create a semaphore for each fork
    for (i = 0; i < PHILOSOPHERS; i++)
    {
        sem_init(&forks[i], 1);
    }

    //  Randomize the order the philosophers are created in
    int order[PHILOSOPHERS];
    for (i = 0; i < PHILOSOPHERS; i++)
        order[i] = i;

    for (i = 0; i < 20; i++)
    {
        int s1 = rand() % PHILOSOPHERS;
        int s2 = rand() % PHILOSOPHERS;
        int tmp = order[s2];
        order[s2] = order[s1];
        order[s1] = tmp;
    }

    for (i = 0; i < PHILOSOPHERS; i++)
    {
        t_create(philosopher, order[i], 1);
    }

    //  Wait for all philosophers to eat
    while (eaten != PHILOSOPHERS)
        t_yield();

    //  Clean up all data
    for (i = 0; i < PHILOSOPHERS; i++)
    {
        sem_destroy(&forks[i]);
    }
    sem_destroy(&lock);
    t_shutdown();
}