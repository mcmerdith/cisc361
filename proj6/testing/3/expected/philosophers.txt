[cisc361:/usa/cshen/361/PA_6/Phase_3/UD_Thread_Sem 1083] ./philosophers
Philosopher 1 is hungry
Philosopher 1 has picked up left  fork 2
Philosopher 1 has picked up right fork 1
Philosopher 4 is hungry
Philosopher 4 has picked up right fork 4
Philosopher 4 has picked up left  fork 0
Philosopher 3 is hungry
Philosopher 0 is hungry
Philosopher 2 is hungry
Philosopher 1 is still eating
Philosopher 4 is still eating
Philosopher 1 is still eating
Philosopher 4 is still eating
Philosopher 1 is still eating
Philosopher 1 has eaten
Philosopher 4 is still eating
Philosopher 4 has eaten
Philosopher 2 has picked up right fork 2
Philosopher 2 has picked up left  fork 3
Philosopher 3 has picked up left  fork 4
Philosopher 0 has picked up right fork 0
Philosopher 0 has picked up left  fork 1
Philosopher 2 is still eating
Philosopher 0 is still eating
Philosopher 2 is still eating
Philosopher 0 is still eating
Philosopher 2 is still eating
Philosopher 2 has eaten
Philosopher 0 is still eating
Philosopher 0 has eaten
Philosopher 3 has picked up right fork 3
Philosopher 3 is still eating
Philosopher 3 is still eating
Philosopher 3 is still eating
Philosopher 3 has eaten
[cisc361:/usa/cshen/361/PA_6/Phase_3/UD_Thread_Sem 1084] ./philosophers
Philosopher 2 is hungry
Philosopher 2 has picked up right fork 2
Philosopher 2 has picked up left  fork 3
Philosopher 1 is hungry
Philosopher 3 is hungry
Philosopher 3 has picked up left  fork 4
Philosopher 4 is hungry
Philosopher 0 is hungry
Philosopher 0 has picked up right fork 0
Philosopher 0 has picked up left  fork 1
Philosopher 2 is still eating
Philosopher 0 is still eating
Philosopher 2 is still eating
Philosopher 0 is still eating
Philosopher 2 is still eating
Philosopher 2 has eaten
Philosopher 0 is still eating
Philosopher 0 has eaten
Philosopher 1 has picked up left  fork 2
Philosopher 1 has picked up right fork 1
Philosopher 3 has picked up right fork 3
Philosopher 1 is still eating
Philosopher 3 is still eating
Philosopher 1 is still eating
Philosopher 3 is still eating
Philosopher 1 is still eating
Philosopher 1 has eaten
Philosopher 3 is still eating
Philosopher 3 has eaten
Philosopher 4 has picked up right fork 4
Philosopher 4 has picked up left  fork 0
Philosopher 4 is still eating
Philosopher 4 is still eating
Philosopher 4 is still eating
Philosopher 4 has eaten
[cisc361:/usa/cshen/361/PA_6/Phase_3/UD_Thread_Sem 1085] ./philosophers
Philosopher 4 is hungry
Philosopher 4 has picked up right fork 4
Philosopher 4 has picked up left  fork 0
Philosopher 3 is hungry
Philosopher 2 is hungry
Philosopher 2 has picked up right fork 2
Philosopher 2 has picked up left  fork 3
Philosopher 0 is hungry
Philosopher 1 is hungry
Philosopher 4 is still eating
Philosopher 2 is still eating
Philosopher 4 is still eating
Philosopher 2 is still eating
Philosopher 4 is still eating
Philosopher 4 has eaten
Philosopher 2 is still eating
Philosopher 2 has eaten
Philosopher 3 has picked up left  fork 4
Philosopher 3 has picked up right fork 3
Philosopher 0 has picked up right fork 0
Philosopher 0 has picked up left  fork 1
Philosopher 1 has picked up left  fork 2
Philosopher 3 is still eating
Philosopher 0 is still eating
Philosopher 3 is still eating
Philosopher 0 is still eating
Philosopher 3 is still eating
Philosopher 3 has eaten
Philosopher 0 is still eating
Philosopher 0 has eaten
Philosopher 1 has picked up right fork 1
Philosopher 1 is still eating
Philosopher 1 is still eating
Philosopher 1 is still eating
Philosopher 1 has eaten
[cisc361:/usa/cshen/361/PA_6/Phase_3/UD_Thread_Sem 1086]